src/main/java/com/iota/iri/service/validation/TransactionValidator.java

```diff
@@ -0,0 +1,167 @@
+        }
+    }
+
+    /**
+     * @return the minimal number of trailing 9s that have to be present at the end of the transaction hash
+     * in order to validate that sufficient proof of work has been done
+     */
+    public int getMinWeightMagnitude() {
+        return minWeightMagnitude;
+    }
+
+    /**
+     * Checks that the timestamp of the transaction is below the last global snapshot time
+     * or more than {@value #MAX_TIMESTAMP_FUTURE} seconds in the future, and thus invalid.
+     *
+     * <p>
+     *     First the attachment timestamp (set after performing POW) is checked, and if not available
+     *     the regular timestamp is checked. Genesis transaction will always be valid.
+     * </p>
+     * @param transactionViewModel transaction under test
+     * @return <tt>true</tt> if timestamp is not in valid bounds and {@code transactionViewModel} is not genesis.
+     * Else returns <tt>false</tt>.
+     */
+    private boolean hasInvalidTimestamp(TransactionViewModel transactionViewModel) {
+        // ignore invalid timestamps for transactions that were requested by our node while solidifying a milestone
+        if(transactionRequester.wasTransactionRecentlyRequested(transactionViewModel.getHash())) {
+            return false;
+        }
+
+        if (transactionViewModel.getAttachmentTimestamp() == 0) {
+            return transactionViewModel.getTimestamp() < snapshotProvider.getInitialSnapshot().getTimestamp() && !snapshotProvider.getInitialSnapshot().hasSolidEntryPoint(transactionViewModel.getHash())
+                    || transactionViewModel.getTimestamp() > (System.currentTimeMillis() / 1000) + MAX_TIMESTAMP_FUTURE;
+        }
+        return transactionViewModel.getAttachmentTimestamp() < (snapshotProvider.getInitialSnapshot().getTimestamp() * 1000L)
+                || transactionViewModel.getAttachmentTimestamp() > System.currentTimeMillis() + MAX_TIMESTAMP_FUTURE_MS;
+    }
+
+    /**
+     * Runs the following validation checks on a transaction:
+     * <ol>
+     *     <li>{@link #hasInvalidTimestamp} check.</li>
+     *     <li>Check that no value trits are set beyond the usable index, otherwise we will have values larger
+     *     than max supply.</li>
+     *     <li>Check that sufficient POW was performed.</li>
+     *     <li>In value transactions, we check that the address has 0 set as the last trit. This must be because of the
+     *     conversion between bytes to trits.</li>
+     * </ol>
+     *Exception is thrown upon failure.
+     *
+     * @param transactionViewModel transaction that should be validated
+     * @param minWeightMagnitude the minimal number of trailing 9s at the end of the transaction hash
+     * @throws StaleTimestampException if timestamp check fails
+     * @throws IllegalStateException if any of the other checks fail
+     */
+    public void runValidation(TransactionViewModel transactionViewModel, final int minWeightMagnitude) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'runValidation(TransactionViewModel, int)' has a cyclomatic complexity of 11.](https://app.codacy.com/manual/GalRogozinski/iri/pullRequest?prid=4339340)